java-checks/src/main/java/org/sonar/java/checks/PrivateFieldUsedLocallyCheck.java

```diff
@@ -0,0 +1,155 @@
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import static org.sonar.java.se.ProgramState.isField;
+
+@Rule(key = "S1450")
+public class PrivateFieldUsedLocallyCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE = "Remove the \"%s\" field and declare it as a local variable in the relevant methods.";
+
+  private Map<MethodTree, Set<Symbol>> fieldsLiveInMethodEntryPerMethod = new HashMap<>();
+  private Set<Symbol> fieldsReadOnAnotherInstance = new HashSet<>();
+
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    TypeSymbol classSymbol = ((ClassTree) tree).symbol();
+    initFieldsMetaData(classSymbol);
+
+    classSymbol.memberSymbols().stream()
+      .filter(PrivateFieldUsedLocallyCheck::isPrivateField)
+      .filter(memberSymbol -> !memberSymbol.usages().isEmpty())
+      .filter(memberSymbol -> !isUsedOutsideMethods(memberSymbol, classSymbol))
+      .forEach(this::checkPrivateField);
+
+    fieldsReadOnAnotherInstance.clear();
+    fieldsLiveInMethodEntryPerMethod.clear();
+  }
+
+  private static boolean isPrivateField(Symbol memberSymbol) {
+    return memberSymbol.isPrivate() && memberSymbol.isVariableSymbol();
+  }
+
+  private static boolean isUsedOutsideMethods(Symbol privateFieldSymbol, TypeSymbol classSymbol) {
+    for (IdentifierTree usageIdentifier : privateFieldSymbol.usages()) {
+      Tree containingClassOrMethod = containingClassOrMethod(usageIdentifier);
+
+      if (containingClassOrMethod.is(Kind.CLASS)
+        || (containingClassOrMethod.is(Kind.METHOD) && !((MethodTree) containingClassOrMethod).symbol().owner().equals(classSymbol))) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private void checkPrivateField(Symbol privateFieldSymbol) {
+    if (fieldsReadOnAnotherInstance.contains(privateFieldSymbol)) {
+      return;
+    }
+
+    for (Set<Symbol> fieldsLiveInMethodEntry : fieldsLiveInMethodEntryPerMethod.values()) {
+      if (fieldsLiveInMethodEntry.contains(privateFieldSymbol)) {
+        return;
+      }
+    }
+
+    reportIssue(((VariableTree) privateFieldSymbol.declaration()).simpleName(), String.format(MESSAGE, privateFieldSymbol.name()));
+  }
+
+  private static Tree containingClassOrMethod(IdentifierTree usageIdentifier) {
+    Tree parent = usageIdentifier;
+    do {
+      parent = parent.parent();
+    } while (!parent.is(Kind.METHOD, Kind.CLASS));
+
+    return parent;
+  }
+
+  private void initFieldsMetaData(TypeSymbol classSymbol) {
+    fieldsReadOnAnotherInstance = FieldsReadOnAnotherInstanceVisitor.getFrom(classSymbol);
+
+    for (Symbol memberSymbol : classSymbol.memberSymbols()) {
```

Comment:
Styling : This can maybe be expressed nicely using streams 

`classSymbol.memberSymbols().stream().filter(...).map( ms-> ms.declaration()).filter(..).forEach( //build CFG LVA etc);`